File: api/src/health.ts
// Health Check Endpoints
// Story: 0-11 Staging Auto-Deployment
// AC: 4 - Readiness + liveness probes
// AC: 5 - Rollback on failed health checks (Kubernetes uses probe responses)
//
// /health  - Liveness probe: is the process alive?
// /ready   - Readiness probe: is the service ready to accept traffic?
//
// Usage: import { registerHealthRoutes } from './health';
//        registerHealthRoutes(app);

import type { Express, Request, Response } from 'express';

interface HealthResponse {
  status: string;
  timestamp: string;
}

interface ReadyResponse extends HealthResponse {
  checks?: Record<string, string>;
  error?: string;
}

/**
 * Register /health and /ready endpoints on the Express app.
 *
 * @param app - Express application instance
 * @param deps - Optional dependency check functions (database, redis, etc.)
 */
export function registerHealthRoutes(
  app: Express,
  deps?: {
    checkDatabase?: () => Promise<void>;
    checkRedis?: () => Promise<void>;
  },
): void {
  // Liveness probe: returns 200 if the process is running
  app.get('/health', (_req: Request, res: Response) => {
    const body: HealthResponse = {
      status: 'ok',
      timestamp: new Date().toISOString(),
    };
    res.status(200).json(body);
  });

  // Readiness probe: returns 200 only if all dependencies are reachable
  app.get('/ready', async (_req: Request, res: Response) => {
    const checks: Record<string, string> = {};

    try {
      if (deps?.checkDatabase) {
        await deps.checkDatabase();
        checks.database = 'ok';
      }

      if (deps?.checkRedis) {
        await deps.checkRedis();
        checks.redis = 'ok';
      }

      const body: ReadyResponse = {
        status: 'ready',
        timestamp: new Date().toISOString(),
        checks,
      };
      res.status(200).json(body);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      const body: ReadyResponse = {
        status: 'not ready',
        timestamp: new Date().toISOString(),
        checks,
        error: message,
      };
      res.status(503).json(body);
    }
  });
}

File: backend/src/health.py
"""
Health Check Endpoints
Replaces: api/src/health.ts (TypeScript/Express) — Sprint Change Proposal 2026-02-20
Story: 0-11 Staging Auto-Deployment
AC: 4 - Readiness + liveness probes
AC: 5 - Rollback on failed health checks (Kubernetes uses probe responses)

/health  - Liveness probe: is the process alive?
/ready   - Readiness probe: is the service ready to accept traffic?

Usage:
    from src.health import router as health_router
    app.include_router(health_router)
"""

from datetime import datetime, timezone
from typing import Callable, Optional

from fastapi import APIRouter, status
from fastapi.responses import JSONResponse

router = APIRouter(tags=["health"])

# Pluggable dependency check callables — set at app startup via main.py
_check_database: Optional[Callable] = None
_check_redis: Optional[Callable] = None


def register_health_checks(
    check_database: Optional[Callable] = None,
    check_redis: Optional[Callable] = None,
) -> None:
    """Register dependency check callables for the readiness probe."""
    global _check_database, _check_redis
    _check_database = check_database
    _check_redis = check_redis


@router.get("/health")
async def liveness() -> JSONResponse:
    """Liveness probe: returns 200 if the process is running."""
    return JSONResponse(
        status_code=status.HTTP_200_OK,
        content={
            "status": "ok",
            "timestamp": datetime.now(timezone.utc).isoformat(),
        },
    )


@router.get("/ready")
async def readiness() -> JSONResponse:
    """Readiness probe: returns 200 only if all dependencies are reachable."""
    checks: dict[str, str] = {}

    try:
        if _check_database is not None:
            await _check_database()
            checks["database"] = "ok"

        if _check_redis is not None:
            await _check_redis()
            checks["redis"] = "ok"

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "status": "ready",
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "checks": checks,
            },
        )

    except Exception as exc:
        return JSONResponse(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            content={
                "status": "not ready",
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "checks": checks,
                "error": str(exc),
            },
        )
